### PR TITLE
Flow.retry using Schedule

### PR DESCRIPTION
**Pull request by LordRaydenMK**
## Issue
Fixes: https://github.com/arrow-kt/arrow-fx/issues/326

## Status
**READY**

## Description
Retry operator for `Flow` using `Schedule`

## Todos
- [x] Tests
- [x] Documentation

Thanks @nomisRev for your help and support.

### CODE REVIEW

File: arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @see [Schedule] for how to build a schedule.
+ *
+ * ```kotlin:ank:playground
+ * import kotlinx.coroutines.flow.*
+ * import arrow.fx.coroutines.*
+ * suspend fun main(): Unit {
+ *   val flow = flow {
+ *    emit(a)
+ *    if (++counter <= 5) throw RuntimeException("Bang!")
```
Comment (nomisRev) [inline arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt:22]: Where is this `counter` coming from?

File: arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        emit(a)
+        if (++counter <= 5) throw RuntimeException("Bang!")
+      }.retry(Schedule.recurs(n))
+        .reduce { acc, int -> acc + int }
+
+      sum shouldBe a * 6
+    }
+  }
+
+  // Note: this takes 50-60 sec to run
```
Comment (nomisRev) [inline arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt:43]: We should use KotlinX Coroutines Test here: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-blocking-test.html

That should make this test instant, whilst it should also stabalise any flakyness that might occur by doing time measurements on the JVM.

See: https://github.com/arrow-kt/arrow-fx/blob/master/arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/BracketCaseTest.kt#L21

**Review by nomisRev — COMMENTED** [review]
Looks good. Only one concern with the long running test, and the missing `counter` definition in the doc snippet.

File: arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt (hunk 2 above)
Comment (LordRaydenMK) [reply arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt]: Didn't know this existed. Today I learned :smile:

**Review by nomisRev — APPROVED** [review]
Awesome new operator 👏 👏 👏 Thanks @LordRaydenMK!

File: arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      val sum = flow {
+        emit(a)
+        if (++counter <= 5) throw RuntimeException("Bang!")
+      }.retry(Schedule.recurs(n))
+        .reduce { acc, int -> acc + int }
+
+      sum shouldBe a * 6
+    }
+  }
+
```
Comment (aballano) [inline arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt:43]: I would add an additional test to make sure we don't retry on CancellationException

File: arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * @param schedule - the [Schedule] used for retrying the collection of the flow
+ */
+fun <A, B> Flow<A>.retry(schedule: Schedule<Throwable, B>): Flow<A> = flow {
+  (schedule as Schedule.ScheduleImpl<Any?, Throwable, B>)
+  var dec: Schedule.Decision<Any?, B>
+  var state: Any? = schedule.initialState()
+
+  val retryWhen = retryWhen { cause, _ ->
+    dec = schedule.update(cause, state)
```
Comment (aballano) [inline arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt:44]: Shouldn't we check for a possible cancellation exception here and rethrow?

**Review by aballano — COMMENTED** [review]
Looks quite good, only a question regarding cancellation

File: arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt (hunk 4 above)
Comment (LordRaydenMK) [reply arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt:44]: That should be handled by `retryWhen` 

https://github.com/Kotlin/kotlinx.coroutines/blob/22087ef122b5ac677fd14941a18b6aefc57393d2/kotlinx-coroutines-core/common/src/flow/operators/Errors.kt#L207

**Review by aballano — APPROVED** [review]

File: arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt (hunk 4 above)
Comment (aballano) [reply arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/flow.kt:44]: You're right, thank you!

File: arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt (hunk 3 above)
Comment (LordRaydenMK) [reply arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/FlowTest.kt:43]: Since that is handled by `retryWhen` do we really need a test?

**Pull request merged**